controllers/custom/builder.go

```diff
@@ -130,24 +130,30 @@ func (b *Builder) Complete(reconciler Reconciler) (healthz.Checker, error) {
 				if err != nil {
 					return err
 				}
+				metaObj, ok := convertedObj.(metav1.Object)
+				if !ok {
+					return fmt.Errorf("failed to get object meta %v", obj)
+				}
+				b.log.Info("processing item", "objName", metaObj.GetName(), "deltaType", d.Type)
 				switch d.Type {
 				case cache.Sync, cache.Added, cache.Updated:
-					if _, exists, err := b.dataStore.Get(convertedObj); err == nil && exists {
-						if err := b.dataStore.Update(convertedObj); err != nil {
-							return err
+
+					if _, exists, err := b.dataStore.Get(convertedObj); err == nil {
+						if exists {
+							if err := b.dataStore.Update(convertedObj); err != nil {
+								b.log.Error(err, "failed to update object in datastore, returning error", "obj", metaObj.GetName())
+								return err
+							}
+						} else {
+							if err := b.dataStore.Add(convertedObj); err != nil {
+								b.log.Error(err, "failed to add object to datastore, returning error", "obj", metaObj.GetName())
+								return err
+							}
 						}
 					} else {
-						if err := b.dataStore.Add(convertedObj); err != nil {
-							return err
-						}
-					}
-					if err != nil {
```

Comment:
Quick note, `err!=nil` was failing the analyzer nil-ness check in VSCode and flagged as warning as this condition is never executed. 

The logic remains unchanged otherwise. `dataStore.Get` returns error only when `KeyFunc` fails (so should ideally never return an error here).